app/controllers/Application.java

```diff
@@ -1416,4 +1435,129 @@ public static Result restFlowMetricsGraphData(String flowDefId) {
 
     return results;
   }
+
+  public static Result restDagGraphData(String flowExecId) {
+    int i,j,oneInnodeNum;
+    JsonObject datasets = new JsonObject();
+
+    if (flowExecId == null || flowExecId.isEmpty()) {
+      return ok(new Gson().toJson(datasets));
+    }
+
+    int max_index= AppJobNameMap.find.select("*").where().eq(AppJobNameMap.TABLE.FLOW_EXEC_ID, flowExecId).findRowCount();
+    max_index= max_index+1;
+    int[][] adjMatrix = new int[max_index][max_index];
+
+    List<AppJobNameMap> reqList= AppJobNameMap.find.select("*").where().eq(AppJobNameMap.TABLE.FLOW_EXEC_ID, flowExecId).findList();
+    for(AppJobNameMap temp1 : reqList){
+      int ownId= temp1.jobNameId;
+      String innodes= temp1.jobInnodes;
+      if(innodes!=null){
+        String[] innodeString = innodes.split(",");
+        for(String oneInnodeString : innodeString){
+          oneInnodeNum= Integer.parseInt(oneInnodeString);
+          adjMatrix[oneInnodeNum][ownId]= 1;
+        }
+      }
+    }
+
+    for (i = 1; i < max_index; i++) {
+      JsonObject row = new JsonObject();
+      for (j = 1; j < max_index; j++) {
+        if(adjMatrix[i][j]==1){
+          row.addProperty(Integer.toString(j), adjMatrix[i][j]);
+        }
+      }
+      AppJobNameMap tempStrStruct= AppJobNameMap.find.select("*").where().eq(AppJobNameMap.TABLE.FLOW_EXEC_ID, flowExecId).eq(AppJobNameMap.TABLE.JOB_NAME_ID, i).findUnique();
+      String label= tempStrStruct.jobName;
+
+      String colour1= "blue";
+
+      List<AppResult> timeStructList= AppResult.find.select("*").where().eq(AppResult.TABLE.FLOW_EXEC_ID, flowExecId).eq(AppResult.TABLE.JOB_NAME, label).findList();
+      long runningTime = 10;
```

Comment:
Can you move the magic string and integer literals such as blue and 10 to static final contstants?